### PR TITLE
feat(client): list-clients / revoke-client CLI for owner-side cleanup

### LINKED ISSUE #166: CLI: add `list-clients` / `revoke-client` for owner-side client cleanup
## Problem

`vicoop-client` lets an owner **register** a client end-to-end without leaving the CLI (`setup` mints a `client_token`, writes a row into `clients`, registers the agent_id allowlist). The reverse side — listing what's registered under your owner principal and **revoking** stale rows — has no CLI surface today. Operators have to drop to admin GraphQL or raw SQL on the bridge DB, which:

- isn't available to non-admin owners (the SIWE-exchanged owner-session is enough to mint a client, but not to access `psql`),
- breaks the otherwise-self-contained `vicoop-client` story documented in [`docs/install-client.md`](./docs/install-client.md),
- leaves no way to clean up the residue from a typo'd `--client-name`, an aborted setup, an exfiltrated token, or test clients spun up during PR validation.

Concrete repro from #165 review: I registered two test clients (`claude-usage-test-*` and `codex-usage-test-*`) to validate the openai-compat `usage` forwarding against the deployed bridge. After killing both daemons the rows persist in `clients`, and `vicoop-client list-agents` only shows currently-**connected** agents — so there's no way from the CLI to even *see* the orphaned rows, let alone clean them up.

Existing surface for reference:

```
subcommands: login, setup, upgrade, list-agents, list-callers, add-caller, remove-caller, whoami
```

`add-caller` / `remove-caller` already establish the pattern for owner-session-authenticated mutations against agent-policy rows. `list-clients` / `revoke-client` would be the analogous pair for the `clients` rows themselves.

## Proposed CLI surface

```
vicoop-client list-clients [--bridge URL] [--json]
vicoop-client revoke-client <client-id-or-name> [--bridge URL]
```

`list-clients` lists every `clients` row whose `owner_principal` matches the owner-session bound principal. **Tokens are never re-emitted** (only the `token_hash` is stored anyway). Columns: `client_id`, `client_name`, `allowed_agent_ids`, `created_at`, plus a `connected` boolean (current registry membership) so the operator can spot orphans at a glance.

`revoke-client` resolves the argument as either a UUID `client_id` or a unique `client_name`, then deletes the row (or flips a `revoked_at`/`revoked=true` column — see "Soft vs hard" below). When the targeted client has a live WS connection, the server closes it (similar in spirit to the 4009 "replaced by new connection" close code already used on token collision, but with a different code, e.g. 4010 "revoked").

Both subcommands use the same owner-session bearer flow as `add-caller` / `remove-caller`, so an existing `vicoop-client login`-acquired session is sufficient. No SIWE re-sign needed.

## Server side

Either a pair of admin REST endpoints scoped to owner-session, or new mutations on the existing PostGraphile-exposed admin schema (whichever matches the `add-caller` plumbing more closely). Logic:

- `listClientsForOwner(principal)` — SELECT … WHERE `owner_principal = $1` AND (no revoked column or `revoked_at IS NULL`).
- `revokeClient(principal, target)` — UPDATE … SET `revoked_at = now()` (soft) or DELETE (hard); cascade should also clean up `agent_policies` rows for `allowed_agent_ids` that nobody else owns. Atomic in a single transaction.

## Soft vs hard delete

Two viable shapes, pick one explicitly:

| Approach | Pros | Cons |
|---|---|---|
| **Soft** (`revoked_at TIMESTAMP NULL`) | Auditable, reversible during the rotation window, easy to add reason/`revoked_by` columns later, `clients` row stays as historical record. | Schema migration adds a column + every `clients` SELECT path now needs the `AND revoked_at IS NULL` guard. |
| **Hard** (`DELETE FROM clients`) | Zero schema churn, no risk of forgetting the soft-delete predicate on a new query path. | Loses the audit trail; if a revocation was a mistake, the operator has to `setup` again from scratch, which also issues a brand-new token and `client_id`. |

Recommendation: **soft** by default with `--hard` opt-in. Audit trail is cheap and the gateway side will eventually want it for billing reconciliation.

## Edge cases

- **Token cache TTL.** `clients` token verification has an LRU cache (per `docs/local-testing.md` Gotchas, 60s on caller tokens — confirm whether `clients` token verification is similarly cached). Revocation should at minimum document the propagation window; ideally the cache is invalidated synchronously on revoke.
- **Live daemon disconnect.** Revoking a client whose daemon is connected should force-close the WS with a distinct close code so the daemon's reconnect loop knows to stop trying (vs blindly retrying for hours). Suggest 4010 `revoked` — paired with an early-exit branch in `client.ts` reconnect logic so the daemon process surfaces "revoked, exiting" and quits non-zero.
- **`allowed_agent_ids` orphan policies.** When a revoked client was the sole owner of an `agent_policies` row, that policy should be cleaned up in the same transaction (or marked stale) so the agent_id namespace doesn't accumulate dead entries.
- **Name collision on `revoke-client <name>`.** If two of the owner's clients share the same `client_name`, fail loud with "ambiguous, specify client_id" rather than picking arbitrarily.

## Out of scope

- Cross-owner revocation (admin-only). Already covered by admin GraphQL.
- Caller token revocation (`callers` rows). Different table, different flow — file a sibling issue if needed.
- Token rotation in place (revoke + reissue under same `client_id`). Useful but distinct from cleanup; could be a follow-up `rotate-client` subcommand sharing the same auth path.

## Acceptance criteria

- [ ] `vicoop-client list-clients` lists owner-scoped `clients` rows with `client_id`, `client_name`, `allowed_agent_ids`, `created_at`, `connected`.
- [ ] `vicoop-client revoke-client <id-or-name>` deletes/soft-deletes the row, returns exit 0 on success, exit non-zero with a usable message on not-found / ambiguous / permission errors.
- [ ] Live daemon's WS is closed with a distinct close code on revoke; daemon-side reconnect logic stops retrying on that code.
- [ ] `clients`-row revocation propagation window documented (or made synchronous).
- [ ] `docs/install-client.md` updated to point at the new subcommands instead of "drop into admin GraphQL / SQL."
- [ ] Unit tests in `packages/client/src/*.test.ts` cover the auth matrix (no session / wrong principal / valid principal) and the name-resolution edge cases.

Refs #165 (the PR that prompted this — two test clients are still hanging around because there's no way to take them out from the CLI).

### PR DESCRIPTION
**Pull request by longfin**
Closes #166.

## Summary

- `vicoop-client list-clients` — lists every `clients` row owned by the operator with a `connected` flag so orphans left behind by an aborted setup or an exited daemon are visible from the CLI.
- `vicoop-client revoke-client <id-or-name>` — sets `revoked = true` on the row, closes any live WS bound to that client, and tells the daemon to exit. Resolves UUID `client_id` or unique `client_name`; ambiguous names return 409 with candidate ids.
- New WS close code **4014 "client revoked"** + daemon-side branch that exits non-zero without reconnecting.

## Design notes

- **No schema migration.** `clients.revoked BOOLEAN` and `revoke_client(TEXT)` already exist in `schema.sql`. Promoting to `revoked_at TIMESTAMPTZ` is filed as a follow-up — orthogonal to the CLI surface this PR ships and has a much larger blast radius (`client_with_token` TYPE, admin LLM prompt, every `WHERE revoked = false` predicate).
- **Close code 4014.** Audited the bridge's WS close-code range before picking: 4001-4008 ws.ts, 4009 registry.ts, 4010-4011 ws.ts, 4012-4013 card-resolver.ts. 4014 is the next free slot; the rationale lives in `Registry.disconnectClient` so future close-code choices stay collision-free. (Earlier revisions of this PR tried 4010 — collided with `agent id owned by a different principal` — and then 4012 — collided with `missing agent card or backend kind`. Both were caught and fixed in review.)
- **`process.exit` lifted out of the Client class.** The daemon's fatal close handler now invokes a new `ClientOptions.onFatal` callback; the Client only tears down its own state (`stopped = true`, reconnect timer cleared, inflight tasks aborted). The production entrypoint in `cli.ts` wires `onFatal: () => process.exit(1)`; tests pass capturing callbacks. This keeps the class testable and friendly to future in-process embedders.
- **No cache invalidation needed.** `ws.ts:27` already queries `clients` directly on every WS register — there is no equivalent of the 60s `callers` LRU cache on client tokens, so revocation is synchronous from the next auth attempt. Documented as such in `install-client.md`.
- **RLS gives the 404 / 409 / cross-owner story for free.** `listClientsForOwner` and `revokeClientForOwner` go through `setRlsContext()` like every other admin-api function. A name that resolves uniquely under the operator's RLS view is unambiguous; one that matches multiple rows triggers a deliberate 409 with the candidate ids so the operator can retry with the id.

## Files

- `packages/server/src/admin-api.ts` — `listClientsForOwner`, `revokeClientForOwner`, `resolveClient` (handles id vs unique-name vs ambiguous-name).
- `packages/server/src/registry.ts` — `disconnectClient(clientId): number`, closes every matching ws with 4014; includes the full close-code audit comment.
- `packages/server/src/http.tsx` — `GET /admin-api/clients`, `DELETE /admin-api/clients/:target`, with 409 added to the AdminApiError status union.
- `packages/client/src/admin-cli.ts` — `runListClients`, `runRevokeClient`, plus renderers.
- `packages/client/src/cli.ts` — subcommand wiring + updated `SUBCOMMAND_LIST` + `onFatal: () => process.exit(1)` for the daemon path.
- `packages/client/src/client.ts` — 4014 branch in the WS close handler, calls `opts.onFatal` instead of `process.exit`.
- `packages/client/src/client.test.ts` — two new regressions: 4014 → onFatal + no reconnect, 1012 → reconnect + no onFatal.
- `docs/install-client.md` — new "Inspecting and revoking your clients" section.
- `.changeset/list-revoke-client-cli.md` — `@vicoop-bridge/client` + `@vicoop-bridge/server` minor.

## Test plan

- [x] `pnpm typecheck` clean
- [x] `packages/client/src/admin-cli.test.ts` — 14 / 14 pass (6 new for list/revoke)
- [x] `packages/client/src/client.test.ts` — 43 / 43 pass (2 new for 4014 / onFatal)
- [x] `packages/server/src/admin-api.test.ts` — 14 / 14 pass (8 new: RLS / 4014 / ambiguous name / orphan listing / unauthorized)
- [x] `packages/server/src/registry.test.ts` — 9 / 9 pass (2 new for `disconnectClient`)
- [x] Full server suite: 208 / 210 pass. The 2 failures are pre-existing in `src/auth/siwe-exchange.test.ts` (audience-mismatch, reproduced on clean main); unrelated to this change.
- [ ] Manual: `vicoop-client list-clients` + `revoke-client` against a local bridge with a live daemon to eyeball the human renderer and confirm the daemon exits on 4014.

## Out of scope (per the issue)

- Cross-owner revocation (admin-only; covered by admin GraphQL)
- Caller token revocation (different table, different flow)
- In-place token rotation (`rotate-client` — already exists at the GraphQL layer)
- `revoked_at TIMESTAMPTZ` audit-trail upgrade (separate follow-up)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/server/src/registry.ts [hunk 1]
```diff
@@ -73,6 +73,28 @@ export class Registry {
...
+  // which in turn calls unregisterAgent and removes the entry from the
+  // `agents` map and flushes any bound tasks via the usual disconnect
+  // path. We don't pre-delete from `agents` here — letting the close
+  // handler do it keeps the in-memory state machine single-sourced.
+  disconnectClient(clientId: string): number {
+    let closed = 0;
+    for (const conn of this.agents.values()) {
+      if (conn.clientId !== clientId) continue;
+      conn.ws.close(4012, 'client revoked');
+      closed++;
```
[the remark this answers is not in the document]
Comment (longfin) [reply packages/server/src/registry.ts:105]: Fixed in 4fa66e6. Audited the close-code range: 4001-4013 are all claimed (ws.ts 4001-4011, card-resolver.ts 4012-4013). Swapped to 4014 "client revoked" and added a comment in `Registry.disconnectClient` enumerating the occupied slots so future close-code choices stay collision-free.

File: packages/client/src/client.ts [hunk 2]
```diff
@@ -312,10 +312,23 @@ export class Client {
...
+      // systemd / a parent supervisor surfaces the revocation instead of
+      // masking it as a transient network hiccup.
+      if (code === 4012) {
+        this.logger.error('client revoked by owner; exiting');
+        this.stopped = true;
+        this.clearReconnectTimer();
+        for (const controller of this.inflight.values()) controller.abort();
+        process.exit(1);
       }
+      this.scheduleReconnect();
```
[the remark this answers is not in the document]
Comment (longfin) [reply packages/client/src/client.ts]: Fixed in 4fa66e6 along with the broader 4012 sweep — daemon now branches on 4014. The companion `process.exit` refactor in the sibling comment also helps here: even if a future regression resurrected the ambiguity, the `onFatal` callback makes the fatal exit observable in tests so the misclassification would fail loudly instead of silently `exit(1)`-ing the test runner.

File: packages/client/src/client.ts [hunk 3]
```diff
@@ -312,10 +312,23 @@ export class Client {
...
+      // same token. Looping forever would just spam reconnects against a
+      // permanently-failing auth; stop the daemon with a non-zero exit so
+      // systemd / a parent supervisor surfaces the revocation instead of
+      // masking it as a transient network hiccup.
+      if (code === 4012) {
+        this.logger.error('client revoked by owner; exiting');
+        this.stopped = true;
+        this.clearReconnectTimer();
+        for (const controller of this.inflight.values()) controller.abort();
+        process.exit(1);
```
[the remark this answers is not in the document]
Comment (longfin) [reply packages/client/src/client.ts]: Fixed in 4fa66e6. Replaced the inline `process.exit(1)` with a new `ClientOptions.onFatal` callback. The Client only tears down its own state (`stopped = true`, reconnect timer cleared, inflight aborted) and delegates the process-lifecycle decision to the caller. cli.ts wires `onFatal: () => process.exit(1)` for the daemon entrypoint; tests pass capturing callbacks. Added two regression tests asserting (a) 4014 triggers onFatal exactly once with no reconnect, (b) other close codes (1012) reconnect without firing onFatal.

File: packages/server/src/http.tsx [hunk 4]
```diff
@@ -373,8 +375,11 @@ export function createHttpApp(opts: ServerHttpOptions): Hono {
   function adminApiErrorResponse(c: Context, err: unknown): Response {
     if (err instanceof AdminApiError) {
       // Cast to a Hono-acceptable status union; AdminApiError uses standard
-      // HTTP codes (400/403/404) that Hono accepts for c.json's second arg.
-      return c.json({ error: err.message }, err.status as 400 | 401 | 403 | 404);
+      // HTTP codes (400/403/404/409) that Hono accepts for c.json's second
+      // arg. 409 is the ambiguous-client-name signal from
+      // revokeClientForOwner — the operator passed a name that matches
+      // multiple rows and needs to retry with the id.
+      return c.json({ error: err.message }, err.status as 400 | 401 | 403 | 404 | 409);
```
[the remark this answers is not in the document]
Comment (longfin) [reply packages/server/src/http.tsx:384]: Fixed in 4fa66e6. Expanded the comment to enumerate each code: 400 invalid input, 401 auth (kept in the cast union for a future AdminApiError with that status, not currently emitted via this path), 403 forbidden, 404 not-found / RLS-hidden, 409 ambiguous client name.

File: packages/server/src/http.tsx [hunk 5]
```diff
@@ -425,6 +430,43 @@ export function createHttpApp(opts: ServerHttpOptions): Hono {
...
+    } catch (err) {
+      return adminApiErrorResponse(c, err);
+    }
+  });
+
+  // Revoke a client by id or unique name. The target travels in the URL path
+  // and may be a UUID `client_id` or a `client_name`; the server resolves
+  // the ambiguity (404 if neither matches, 409 with the candidate ids if a
+  // name matches multiple rows). On success, sets `revoked = true` and
+  // closes every live WS bound to the client with code 4012.
```
[the remark this answers is not in the document]
Comment (longfin) [reply packages/server/src/http.tsx]: Fixed in 4fa66e6. Route comment now references 4014 and points at the close-code rationale in `Registry.disconnectClient`, where the full audit lives.

File: docs/install-client.md [hunk 6]
```diff
@@ -753,8 +753,49 @@ section of `docs/local-testing.md` for the same note.
...
+  survives; existing `agent_policies` cascade-deleted only when the
+  underlying row is later hard-deleted.
+- A unique name resolves automatically; an ambiguous name exits non-zero
+  with a list of matching `client_id`s so you can retry with the id.
+- If the daemon is alive at the moment of revocation, its WebSocket is
+  closed with code **4012 "client revoked"** and the daemon exits
+  non-zero without reconnecting. The bridge's WS auth path also rejects
+  the token on the next register attempt, so a daemon launched again with
+  the same token will fail with **close code 4005 "bad token"** rather
+  than reconnect-loop indefinitely.
```
[the remark this answers is not in the document]
Comment (longfin) [reply docs/install-client.md]: Fixed in 4fa66e6. Docs now state 4014, and the next sentence still reminds operators that a daemon relaunched with the same token after revoke will fail with the regular 4005 "bad token" close instead.

File: .changeset/list-revoke-client-cli.md [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+- Both subcommands authenticate with the existing `vicoop-client login` owner-session bearer — same flow as `add-caller` / `remove-caller`.
+
+**Server surface**
+
+- `GET /admin-api/clients` and `DELETE /admin-api/clients/:target` under the same owner-session bearer guard as the existing `/admin-api/agents/*` routes. RLS filters list/delete to the operator's own rows; reads of another principal's rows return 404 (no existence leak), name-resolution ambiguity returns 409.
+- `Registry.disconnectClient(clientId)` closes every live WebSocket bound to a revoked client with new close code **4012 "client revoked"**. (4010 was already taken by the agent-id-owned-by-different-principal path in `ws.ts`.)
+
+**Daemon behavior**
+
+- The client daemon's reconnect loop now branches on close code 4012: log `client revoked by owner; exiting`, abort inflight tasks, and exit non-zero without reconnecting. Without this branch the daemon would loop forever against a permanently-failing auth.
```
[the remark this answers is not in the document]
Comment (longfin) [reply .changeset/list-revoke-client-cli.md]: Fixed in 4fa66e6 — changeset now references 4014 to match the wire behavior.

File: packages/server/src/registry.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (longfin) [reply packages/server/src/registry.ts:105]: Already addressed in 4fa66e6 (swap to 4014, `onFatal` callback refactor, 401 comment fix). My initial `git push` of the fix silently failed because the worktree branch had lost its upstream — Copilot re-reviewed the original 8ca8bde, so this round of comments duplicates round 1. The fix commit is now on origin; please re-review against 4fa66e6.

File: packages/client/src/client.ts [hunk 8]
```diff
@@ -312,10 +312,23 @@ export class Client {
...
+      // 4012 is the server's "client revoked" signal (issue #166). The DB
+      // row was just soft-deleted by an owner-side `vicoop-client
+      // revoke-client`, and ws.ts will reject every future hello with the
+      // same token. Looping forever would just spam reconnects against a
+      // permanently-failing auth; stop the daemon with a non-zero exit so
+      // systemd / a parent supervisor surfaces the revocation instead of
+      // masking it as a transient network hiccup.
+      if (code === 4012) {
+        this.logger.error('client revoked by owner; exiting');
+        this.stopped = true;
```
[the remark this answers is not in the document]
Comment (longfin) [reply packages/client/src/client.ts]: Already addressed in 4fa66e6 (swap to 4014, `onFatal` callback refactor, 401 comment fix). My initial `git push` of the fix silently failed because the worktree branch had lost its upstream — Copilot re-reviewed the original 8ca8bde, so this round of comments duplicates round 1. The fix commit is now on origin; please re-review against 4fa66e6.

File: packages/client/src/client.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (longfin) [reply packages/client/src/client.ts]: Already addressed in 4fa66e6 (swap to 4014, `onFatal` callback refactor, 401 comment fix). My initial `git push` of the fix silently failed because the worktree branch had lost its upstream — Copilot re-reviewed the original 8ca8bde, so this round of comments duplicates round 1. The fix commit is now on origin; please re-review against 4fa66e6.

File: packages/server/src/http.tsx [hunk 9]
```diff
@@ -373,8 +375,11 @@ export function createHttpApp(opts: ServerHttpOptions): Hono {
   function adminApiErrorResponse(c: Context, err: unknown): Response {
     if (err instanceof AdminApiError) {
       // Cast to a Hono-acceptable status union; AdminApiError uses standard
-      // HTTP codes (400/403/404) that Hono accepts for c.json's second arg.
-      return c.json({ error: err.message }, err.status as 400 | 401 | 403 | 404);
+      // HTTP codes (400/403/404/409) that Hono accepts for c.json's second
+      // arg. 409 is the ambiguous-client-name signal from
```
[the remark this answers is not in the document]
Comment (longfin) [reply packages/server/src/http.tsx]: Already addressed in 4fa66e6 (swap to 4014, `onFatal` callback refactor, 401 comment fix). My initial `git push` of the fix silently failed because the worktree branch had lost its upstream — Copilot re-reviewed the original 8ca8bde, so this round of comments duplicates round 1. The fix commit is now on origin; please re-review against 4fa66e6.

File: packages/server/src/http.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (longfin) [reply packages/server/src/http.tsx]: Already addressed in 4fa66e6 (swap to 4014, `onFatal` callback refactor, 401 comment fix). My initial `git push` of the fix silently failed because the worktree branch had lost its upstream — Copilot re-reviewed the original 8ca8bde, so this round of comments duplicates round 1. The fix commit is now on origin; please re-review against 4fa66e6.

File: packages/server/src/admin-api.ts [hunk 10]
```diff
@@ -249,3 +249,133 @@ export function listActiveAgents(
...
+    // revoke_client() is SECURITY INVOKER and re-checks RLS; resolveClient()
+    // already verified the row is visible to this principal so the UPDATE
+    // inside will succeed (or NOT FOUND if a concurrent delete happened, in
+    // which case the function raises — we let that propagate as 500 because
+    // it's a genuinely unexpected race, not a user-facing error condition).
+    await tx`SELECT revoke_client(${resolved.id})`;
+  });
+
+  // Close every live WebSocket session bound to this client. The daemon
+  // sees close code 4012 and exits without reconnecting (see client.ts).
```
[the remark this answers is not in the document]
Comment (longfin) [reply packages/server/src/admin-api.ts]: Already addressed in 4fa66e6 (swap to 4014, `onFatal` callback refactor, 401 comment fix). My initial `git push` of the fix silently failed because the worktree branch had lost its upstream — Copilot re-reviewed the original 8ca8bde, so this round of comments duplicates round 1. The fix commit is now on origin; please re-review against 4fa66e6.

File: docs/install-client.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (longfin) [reply docs/install-client.md]: Already addressed in 4fa66e6 (swap to 4014, `onFatal` callback refactor, 401 comment fix). My initial `git push` of the fix silently failed because the worktree branch had lost its upstream — Copilot re-reviewed the original 8ca8bde, so this round of comments duplicates round 1. The fix commit is now on origin; please re-review against 4fa66e6.

File: .changeset/list-revoke-client-cli.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (longfin) [reply .changeset/list-revoke-client-cli.md]: Already addressed in 4fa66e6 (swap to 4014, `onFatal` callback refactor, 401 comment fix). My initial `git push` of the fix silently failed because the worktree branch had lost its upstream — Copilot re-reviewed the original 8ca8bde, so this round of comments duplicates round 1. The fix commit is now on origin; please re-review against 4fa66e6.

File: packages/server/src/registry.ts [hunk 11]
```diff
@@ -73,6 +73,37 @@ export class Registry {
...
 
+  // Close every live WebSocket whose ClientConnection.clientId matches.
+  // Used by the admin-api revoke-client path so a daemon whose client row
+  // was just revoked sees a distinct close code (4014) and exits without
+  // reconnecting. Returns the number of connections closed; callers
+  // surface that to the operator as confirmation that an orphan vs a
+  // live daemon was revoked.
+  //
+  // 4014 is the next unused slot in the bridge's WS close-code range:
+  // 4001-4011 are claimed by ws.ts (hello timeout / frame parse /
```
[the remark this answers is not in the document]
Comment (longfin) [reply packages/server/src/registry.ts]: Fixed in 5c62e74. The audit comment now breaks the range down by actual emitter: 4001-4008 ws.ts, **4009 registry.ts (this file)**, 4010-4011 ws.ts, 4012-4013 card-resolver.ts.

File: docs/install-client.md [hunk 12]
```diff
@@ -753,8 +753,49 @@ section of `docs/local-testing.md` for the same note.
...
+
+- A revoked client's row is kept (`revoked = true`) so audit history
+  survives; existing `agent_policies` cascade-deleted only when the
+  underlying row is later hard-deleted.
+- A unique name resolves automatically; an ambiguous name exits non-zero
+  with a list of matching `client_id`s so you can retry with the id.
+- If the daemon is alive at the moment of revocation, its WebSocket is
+  closed with code **4014 "client revoked"** and the daemon exits
+  non-zero without reconnecting. The bridge's WS auth path also rejects
+  the token on the next register attempt, so a daemon launched again with
```
[the remark this answers is not in the document]
Comment (longfin) [reply docs/install-client.md]: PR description updated via `gh pr edit` — every 4012 reference now reads 4014, and the design-notes section walks through the close-code audit and the prior collisions (4010 → 4012 → 4014) so future readers see the history.

File: packages/client/src/client.ts [hunk 13]
```diff
@@ -312,10 +324,26 @@ export class Client {
...
+      // transient network hiccup).
+      if (code === 4014) {
+        this.logger.error('client revoked by owner; stopping');
+        this.stopped = true;
+        this.clearReconnectTimer();
+        for (const controller of this.inflight.values()) controller.abort();
+        this.opts.onFatal?.({ code, reason: reason.toString() });
+        return;
       }
+      this.scheduleReconnect();
```
[the remark this answers is not in the document]
Comment (longfin) [reply packages/client/src/client.ts]: Fixed in caf2f28. Made the docs claim true: client now treats `code === 4014 || code === 4005` as terminal. New regression test asserts 4005 → onFatal + no reconnect; the old "non-4014 reconnects" test was renamed to make it clear it covers the non-terminal companion for both codes. Docs and changeset updated to describe the unified rule (4014 = live disconnect, 4005 = relaunch / wrong token).

File: docs/install-client.md [hunk 14]
```diff
@@ -753,8 +753,49 @@ section of `docs/local-testing.md` for the same note.
...
+  the same token will fail with **close code 4005 "bad token"** rather
+  than reconnect-loop indefinitely.
```
[the remark this answers is not in the document]
Comment (longfin) [reply docs/install-client.md]: Fixed in caf2f28. Made the docs claim true: client now treats `code === 4014 || code === 4005` as terminal. New regression test asserts 4005 → onFatal + no reconnect; the old "non-4014 reconnects" test was renamed to make it clear it covers the non-terminal companion for both codes. Docs and changeset updated to describe the unified rule (4014 = live disconnect, 4005 = relaunch / wrong token).

File: .changeset/list-revoke-client-cli.md [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+- `Registry.disconnectClient(clientId)` closes every live WebSocket bound to a revoked client with new close code **4014 "client revoked"**. (4010 was already taken by the agent-id-owned-by-different-principal path in `ws.ts`.)
+
+**Daemon behavior**
+
+- The client daemon's reconnect loop now branches on close code 4014: log `client revoked by owner; exiting`, abort inflight tasks, and exit non-zero without reconnecting. Without this branch the daemon would loop forever against a permanently-failing auth.
+
+**Revocation propagation**
+
+- Client-token verification in `ws.ts` queries `clients` directly on every WS register (no LRU cache, unlike the 60s `callers` cache documented in `local-testing.md`), so revocation is effectively synchronous from the next auth attempt — a daemon relaunched with the same token after revoke fails with close code 4005 "bad token". No cache-invalidation work needed on the server side.
```
[the remark this answers is not in the document]
Comment (longfin) [reply .changeset/list-revoke-client-cli.md]: Fixed in caf2f28. Made the docs claim true: client now treats `code === 4014 || code === 4005` as terminal. New regression test asserts 4005 → onFatal + no reconnect; the old "non-4014 reconnects" test was renamed to make it clear it covers the non-terminal companion for both codes. Docs and changeset updated to describe the unified rule (4014 = live disconnect, 4005 = relaunch / wrong token).

File: packages/client/src/client.ts [hunk 16]
```diff
@@ -312,10 +325,42 @@ export class Client {
...
-        this.ws = null;
-        this.scheduleReconnect();
+      if (!current) return;
+      this.ws = null;
+      // Terminal auth failures the daemon cannot recover from by waiting:
+      //
+      //   - **4014 "client revoked"** (issue #166). The DB row was just
+      //     soft-deleted by an owner-side `vicoop-client revoke-client`.
+      //     Future hellos will be rejected with 4005 anyway, so we exit
+      //     immediately rather than wait one more reconnect cycle.
```
[the remark this answers is not in the document]
Comment (longfin) [reply packages/client/src/client.ts]: Fixed in ad1458c. The comment now reads "flagged `revoked = true` … the row stays — soft revocation, not deletion — but ws.ts's `revoked = false` filter excludes it."
